### PR TITLE
html5 data-* attributes for data()

### PR DESCRIPTION
**Pull request by rvagg**
Sorry for keeping this up, just adding/fixing things as I want them, I hope you don't mind and that this one meets your approval.

Extended `data()` to also read data-\* attributes from elements as well as internal data storage. It's a one-way thing so they are not written back when set with `data(k, v)` (although that could change of course).

If a key doesn't exist in the internal store then see if it exists as an element data-\* attribute, if it does then fetch it from there, put it in the store and return it. Basic conversion is done for booleans and numbers.

If `data()` is called without any arguments then it's interpreted as a 'get all data' and the whole store is returned; this also means fetching all data-\* attributes on the element in one go and putting them into the store. So it can be used to turn all data-\* attributes into an object which is mighty handy.

Mostly as per: http://dev.w3.org/html5/spec/elements.html#embedding-custom-non-visible-data-with-the-data-attributes

Tests included.

### CODE REVIEW

**Comment by ded** [comment]
you do know the original idea of all this was to keep things small. one hope is that you'd need these features, rather than, as you say, wanting them.

**Comment by ded** [comment]
snarkiness aside, it's been a long vacation. i'll have a look. thanks again for your great contributions

**Comment by rvagg** [comment]
I edited my original text and changed to _want_ from _need_ and I can't remember why now. I'll keep in mind that I have to submit code with a greater level of neediness in future.
I thought only 19 extra lines of code was pretty good for a significant feature.

File: tests/tests.html [hunk 1]
```diff
@@ -477,6 +478,26 @@
           ok(checkedCount() == 3, 'all checkboxes are checked!')
         })
 
+        test('data', 3, function() {
+          var el = $('#data-test'), data
+
+          el.data('disco', 'stu')
```
Comment (ded) [inline tests/tests.html:506]: rock on

File: tests/tests.html [hunk 2]
```diff
@@ -477,6 +478,26 @@
...
+
+          ok(el.data('foo') == 'bar', 'read existing data-* from dom')
+          ok(el.data('disco') == 'stu', 'read data set by data(k,v)')
+
+          data = el.data()          
+          ok(
+            data &&
+            data.foo == 'bar' &&
+            data.disco == 'stu' &&
+            data.neverGunna == 'give you up' &&
```
Comment (ded) [inline tests/tests.html:517]: never forget

**Comment by rvagg** [comment]
ping

any more thoughts on this?

**Comment by ded** [comment]
yeah. busy week

**Comment by rvagg** [comment]
No probs. Just wanted to make sure it hadn't fallen out of view.

**Pull request merged**